File: fc-chat/demo/part3/ch4/message-system-add-push/src/main/java/net/prostars/messagesystem/handler/websocket/FetchChannelsRequestHandler.java
package net.prostars.messagesystem.handler.websocket;

import net.prostars.messagesystem.constant.IdKey;
import net.prostars.messagesystem.dto.domain.UserId;
import net.prostars.messagesystem.dto.websocket.inbound.FetchChannelsRequest;
import net.prostars.messagesystem.dto.websocket.outbound.FetchChannelsResponse;
import net.prostars.messagesystem.service.ChannelService;
import net.prostars.messagesystem.service.ClientNotificationService;
import org.springframework.stereotype.Component;
import org.springframework.web.socket.WebSocketSession;

@Component
@SuppressWarnings("unused")
public class FetchChannelsRequestHandler implements BaseRequestHandler<FetchChannelsRequest> {

  private final ChannelService channelService;
  private final ClientNotificationService clientNotificationService;

  public FetchChannelsRequestHandler(
      ChannelService channelService, ClientNotificationService clientNotificationService) {
    this.channelService = channelService;
    this.clientNotificationService = clientNotificationService;
  }

  @Override
  public void handleRequest(WebSocketSession senderSession, FetchChannelsRequest request) {
    UserId senderUserId = (UserId) senderSession.getAttributes().get(IdKey.USER_ID.getValue());

    clientNotificationService.sendMessage(
        senderSession,
        senderUserId,
        new FetchChannelsResponse(channelService.getChannels(senderUserId)));
  }
}

File: fc-chat/demo/part3/with-push/src/main/kotlin/com/yscorp/withpush/messagesystem/handler/websocket/FetchChannelsRequestHandler.kt
package com.yscorp.withpush.messagesystem.handler.websocket

import com.yscorp.withpush.messagesystem.constant.IdKey
import com.yscorp.withpush.messagesystem.dto.domain.UserId
import com.yscorp.withpush.messagesystem.dto.websocket.inbound.FetchChannelsRequest
import com.yscorp.withpush.messagesystem.dto.websocket.outbound.FetchChannelsResponse
import com.yscorp.withpush.messagesystem.service.ChannelService
import com.yscorp.withpush.messagesystem.service.ClientNotificationService
import org.springframework.stereotype.Component
import org.springframework.web.socket.WebSocketSession

@Component
@Suppress("unused")
class FetchChannelsRequestHandler(
    private val channelService: ChannelService,
    private val clientNotificationService: ClientNotificationService
) : BaseRequestHandler<FetchChannelsRequest> {

    override fun handleRequest(senderSession: WebSocketSession, request: FetchChannelsRequest) {
        val senderUserId = senderSession.attributes[IdKey.USER_ID.value] as UserId

        clientNotificationService.sendMessage(
            senderSession,
            senderUserId,
            FetchChannelsResponse(channelService.getChannels(senderUserId))
        )
    }

}
